File: android/src/main/java/com/onlinepayments/sdk/client/reactnative/masking/MaskingModule.kt
/*
 * Do not remove or alter the notices in this preamble.
 *
 * This software is owned by Worldline and may not be be altered, copied, reproduced, republished, uploaded, posted, transmitted or distributed in any way, without the prior written consent of Worldline.
 *
 * Copyright © 2024 Worldline and/or its affiliates.
 *
 * All rights reserved. License grant and user rights and obligations according to the applicable license agreement.
 *
 * Please contact Worldline for questions regarding license and user rights.
 */

package com.onlinepayments.sdk.client.reactnative.masking

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.google.gson.Gson
import com.onlinepayments.sdk.client.reactnative.extensions.letRequestOrReturnError
import com.onlinepayments.sdk.client.reactnative.extensions.resolveString
import com.onlinepayments.sdk.client.reactnative.masking.models.AccountOnFileMaskedValueRequest
import com.onlinepayments.sdk.client.reactnative.masking.models.AccountOnFileCustomMaskedValueRequest
import com.onlinepayments.sdk.client.reactnative.masking.models.PaymentProductFieldMaskRequest
import com.onlinepayments.sdk.client.reactnative.masking.models.PaymentRequestMaskedValueRequest
import com.onlinepayments.sdk.client.reactnative.masking.models.PaymentRequestAllMaskedValuesRequest

class MaskingModule(private val reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  private val gson = Gson()

  companion object {
    const val NAME = "OnlinePaymentsMasking"
  }

  override fun getName(): String {
    return NAME
  }

  @ReactMethod
  fun applyMaskForPaymentProductField(paymentProductFieldMaskRequest: String, promise: Promise) {
    promise.letRequestOrReturnError(gson, paymentProductFieldMaskRequest) { request: PaymentProductFieldMaskRequest ->
      val field = request.field
      val value = request.value

      val maskedValue = field.applyMask(value)
      promise.resolveString(maskedValue)
    }
  }

  @ReactMethod
  fun removeMaskForPaymentProductField(paymentProductFieldMaskRequest: String, promise: Promise) {
    promise.letRequestOrReturnError(gson, paymentProductFieldMaskRequest) { request: PaymentProductFieldMaskRequest ->
      val field = request.field
      val value = request.value

      val unmaskedValue = field.removeMask(value)
      promise.resolveString(unmaskedValue)
    }
  }

  @ReactMethod
  fun maskedValueForPaymentRequest(paymentRequestMaskedValueRequest: String, promise: Promise) {
    promise.letRequestOrReturnError(
      gson,
      paymentRequestMaskedValueRequest
    ) { request: PaymentRequestMaskedValueRequest ->
      val paymentRequest = request.request
      val fieldId = request.fieldId

      val maskedValue = paymentRequest.getMaskedValue(fieldId)
      promise.resolveString(maskedValue)
    }
  }

  @ReactMethod
  fun unmaskedValueForPaymentRequest(paymentRequestMaskedValueRequest: String, promise: Promise) {
    promise.letRequestOrReturnError(
      gson,
      paymentRequestMaskedValueRequest
    ) { request: PaymentRequestMaskedValueRequest ->
      val paymentRequest = request.request
      val fieldId = request.fieldId

      val unmaskedValue = paymentRequest.getUnmaskedValue(fieldId)
      promise.resolveString(unmaskedValue)
    }
  }

  @ReactMethod
  fun allMaskedValuesForPaymentRequest(paymentRequestAllMaskedValuesRequest: String, promise: Promise) {
    promise.letRequestOrReturnError(
      gson,
      paymentRequestAllMaskedValuesRequest
    ) { request: PaymentRequestAllMaskedValuesRequest ->
      val paymentRequest = request.request

      val allMaskedValues = gson.toJson(paymentRequest.getMaskedValues())
      promise.resolveString(allMaskedValues)
    }
  }

  @ReactMethod
  fun allUnmaskedValuesForPaymentRequest(paymentRequestAllMaskedValuesRequest: String, promise: Promise) {
    promise.letRequestOrReturnError(
      gson,
      paymentRequestAllMaskedValuesRequest
    ) { request: PaymentRequestAllMaskedValuesRequest ->
      val paymentRequest = request.request

      val allUnmaskedValues = gson.toJson(paymentRequest.getUnmaskedValues())
      promise.resolveString(allUnmaskedValues)
    }
  }

  @ReactMethod
  fun maskedValueForAccountOnFile(accountOnFileMaskedValueRequest: String, promise: Promise) {
    promise.letRequestOrReturnError(
      gson,
      accountOnFileMaskedValueRequest
    ) { request: AccountOnFileMaskedValueRequest ->
      val accountOnFile = request.accountOnFile
      val fieldId = request.fieldId

      val maskedValue = accountOnFile.getMaskedValue(fieldId)
      promise.resolveString(maskedValue)
    }
  }

  @ReactMethod
  fun customMaskedValueForAccountOnFile(accountOnFileCustomMaskedValueRequest: String, promise: Promise) {
    promise.letRequestOrReturnError(
      gson,
      accountOnFileCustomMaskedValueRequest
    ) { request: AccountOnFileCustomMaskedValueRequest ->
      val accountOnFile = request.accountOnFile
      val fieldId = request.fieldId
      val mask = request.mask

      val customMaskedValue = accountOnFile.getMaskedValue(fieldId, mask)
      promise.resolveString(customMaskedValue)
    }
  }
}

File: ios/Classes/Masking/MaskingModule.swift
/*
 * Do not remove or alter the notices in this preamble.
 *
 * This software is owned by Worldline and may not be be altered, copied, reproduced, republished,
 * uploaded, posted, transmitted or distributed in any way, without the prior written consent of Worldline.
 *
 * Copyright © 2024 Worldline and/or its affiliates.
 *
 * All rights reserved. License grant and user rights and obligations according to the applicable license agreement.
 *
 * Please contact Worldline for questions regarding license and user rights.
 */
 
import OnlinePaymentsKit

@objc(OnlinePaymentsMasking)
internal class MaskingModule: NSObject {

    @objc(applyMaskForPaymentProductField:withResolver:withRejecter:)
    func applyMaskForPaymentProductField(
        paymentProductFieldMaskRequest: NSString,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        guard let request = PaymentProductFieldMaskRequest.getRequestObject(
            for: paymentProductFieldMaskRequest,
            call: "applyMaskForPaymentProductField",
            resolver: resolve
        ) else {
            return
        }

        let field = request.field
        let value = request.value

        let maskedValue = field.applyMask(value: value)
        resolve(maskedValue)
    }

    @objc(removeMaskForPaymentProductField:withResolver:withRejecter:)
    func removeMaskForPaymentProductField(
        paymentProductFieldMaskRequest: NSString,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        guard let request = PaymentProductFieldMaskRequest.getRequestObject(
            for: paymentProductFieldMaskRequest,
            call: "removeMaskForPaymentProductField",
            resolver: resolve
        ) else {
            return
        }

        let field = request.field
        let value = request.value

        let unmaskedValue = field.removeMask(value: value)
        resolve(unmaskedValue)
    }

    @objc(maskedValueForPaymentRequest:withResolver:withRejecter:)
    func maskedValueForPaymentRequest(
        paymentRequestMaskedValueRequest: NSString,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        guard let request = PaymentRequestMaskedValueRequest.getRequestObject(
            for: paymentRequestMaskedValueRequest,
            call: "maskedValueForPaymentRequest",
            resolver: resolve
        ) else {
            return
        }

        let paymentRequest = request.request
        let fieldId = request.fieldId

        let maskedValue = paymentRequest.maskedValue(forField: fieldId)
        resolve(maskedValue)
    }

    @objc(unmaskedValueForPaymentRequest:withResolver:withRejecter:)
    func unmaskedValueForPaymentRequest(
        paymentRequestUnmaskedValueRequest: NSString,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        guard let request = PaymentRequestMaskedValueRequest.getRequestObject(
            for: paymentRequestUnmaskedValueRequest,
            call: "unmaskedValueForPaymentRequest",
            resolver: resolve
        ) else {
            return
        }

        let paymentRequest = request.request
        let fieldId = request.fieldId

        let unmaskedValue = paymentRequest.unmaskedValue(forField: fieldId)
        resolve(unmaskedValue)
    }

    @objc(allMaskedValuesForPaymentRequest:withResolver:withRejecter:)
    func allMaskedValuesForPaymentRequest(
        paymentRequestAllMaskedValuesRequest: NSString,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        guard let request = PaymentRequestAllMaskedValuesRequest.getRequestObject(
            for: paymentRequestAllMaskedValuesRequest,
            call: "allMaskedValuesForPaymentRequest",
            resolver: resolve
        ) else {
            return
        }

        let paymentRequest = request.request

        guard let allMaskedValues = paymentRequest.maskedFieldValues else {
            resolve(nil)
            return
        }

        allMaskedValues.resolveResult(using: resolve)
    }

    @objc(allUnmaskedValuesForPaymentRequest:withResolver:withRejecter:)
    func allUnmaskedValuesForPaymentRequest(
        paymentRequestAllUnmaskedValuesRequest: NSString,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        guard let request = PaymentRequestAllMaskedValuesRequest.getRequestObject(
            for: paymentRequestAllUnmaskedValuesRequest,
            call: "allUnmaskedValuesForPaymentRequest",
            resolver: resolve
        ) else {
            return
        }

        let paymentRequest = request.request

        guard let allUnmaskedValues = paymentRequest.unmaskedFieldValues else {
            resolve(nil)
            return
        }

        allUnmaskedValues.resolveResult(using: resolve)
    }

    @objc(maskedValueForAccountOnFile:withResolver:withRejecter:)
    func maskedValueForAccountOnFile(
        accountOnFileMaskedValueRequest: NSString,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        guard let request = AccountOnFileMaskedValueRequest.getRequestObject(
            for: accountOnFileMaskedValueRequest,
            call: "maskedValueForAccountOnFile",
            resolver: resolve
        ) else {
            return
        }

        let accountOnFile = request.accountOnFile
        let fieldId = request.fieldId

        let maskedValue = accountOnFile.maskedValue(forField: fieldId)
        resolve(maskedValue)
    }

    @objc(customMaskedValueForAccountOnFile:withResolver:withRejecter:)
    func customMaskedValueForAccountOnFile(
        accountOnFileCustomMaskedValueRequest: NSString,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        guard let request = AccountOnFileCustomMaskedValueRequest.getRequestObject(
            for: accountOnFileCustomMaskedValueRequest,
            call: "customMaskedValueForAccountOnFile",
            resolver: resolve
        ) else {
            return
        }

        let accountOnFile = request.accountOnFile
        let fieldId = request.fieldId
        let mask = request.mask

        let customMaskedValue = accountOnFile.maskedValue(forField: fieldId, mask: mask)
        resolve(customMaskedValue)
    }
}
